# DS 3000 - Assignment [2]

**Student Name**: [Yanzhen Chen]

**Date**: [9/12]


### Submission Instructions
Submit this `ipynb` file to canvas.

The `ipynb` format stores outputs from the last time you ran the notebook.  (When you open a notebook it has the figures and outputs of the last time you ran it too).  To ensure that your submitted `ipynb` file represents your latest code, make sure to give a fresh run `Kernel > Restart & Run All` just before uploading the `ipynb` file to Canvas.

### Academic Integrity

**Writing your homework is an individual effort.**  You may discuss general python problems with other students but under no circumstances should you observe another student's code which was written for this assignment, from this year or past years.  Pop into office hours or DM us in MS Teams if you have a specific question about your work or if you would like another pair of eyes or talk through your code.

Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment. You do not need to cite the official python documentation.

**Documentation / style counts for credit**  Please refer to the Pep-8 style, to improve the readability and consistency of your Python code. For more information, read the following article [How to Write Beautiful Python Code With PEP 8](https://realpython.com/python-pep8/) or ask your TA's for tips.

**NOTE:<span style='color:red'> Write python expressions to answer ALL questions below and ensure that you use the `print()` function to display the output.</span>** Each question should be answered in a new code cell. For example, your solution for question 1.1 should be in a different code cell from your solution for question 1.2.

## Question 1: 

The data that you are working with pertains to the risk factors that are associated with low infant birth weight**. The data contains the following fields:

low
indicator of birth weight less than 2.5 kg.

age
mother's age in years.

lwt
mother's weight in pounds at last menstrual period.

race
mother's race (1 = white, 2 = black, 3 = other).

smoke
smoking status during pregnancy  (0 = false, 1 = true).

ptl
number of previous premature labors.

ht
history of hypertension (0 = false, 1 = true).

ui
presence of uterine irritability.

ftv
number of physician visits during the first trimester.

bwt
birth weight in grams.

Write python code to answer the questions below and ensure that you round all numeric calculations (i.e. floats) to 2 decimal places (unless indicated otherwise).
Tip: you will need to filter for conditions that are true before performing certain calculations.

1. (5 pts) Load the attached data into a numpy array and display the dimensions of the data: birthwt.csvDownload birthwt.csv

In [4]:
import numpy as np

data = np.genfromtxt('birthwt.csv', delimiter=',', skip_header=1)

print("Data dimensions:", data.shape)

Data dimensions: (189, 10)


2. (3 pts) What is the age range of mothers in this dataset.
Note: the range should include: the minimum, average and maximum age. Ensure that the ages are represented as whole numbers.

In [6]:
ages = data[:, 1]

min_age = int(np.min(ages))
avg_age = int(np.mean(ages))
max_age = int(np.max(ages))

print(f"Mother's age range: Min: {min_age}, Avg: {avg_age}, Max: {max_age}")


Mother's age range: Min: 14, Avg: 23, Max: 45


3. (2 pts) How many mothers smoked during pregnancy?

In [8]:
smoke = data[:, 4]

smoking_mothers = np.sum(smoke == 1)
print(f"Number of mothers who smoked during pregnancy: {smoking_mothers}")


Number of mothers who smoked during pregnancy: 74


4. (2 pts) How many mothers had a history of hypertension?

In [10]:
hypertension = data[:, 6]

hypertensive_mothers = np.sum(hypertension == 1)
print(f"Number of mothers with hypertension: {hypertensive_mothers}")


Number of mothers with hypertension: 12


5. (3 pts) How many mothers have 1 or more premature births?

In [12]:
ptl = data[:, 5]

premature_birth_mothers = np.sum(ptl >= 1)
print(f"Number of mothers with 1 or more premature births: {premature_birth_mothers}")


Number of mothers with 1 or more premature births: 30


6. (5 pts) Create a function called convert_birthweight() that: 1) takes grams as input, 2) converts grams to pounds and ounces, and 3) 
return the converted value. Use the convert_birthweight() function to convert the birth weight to pounds and ounces in the questions 7-11 
below.

In [14]:
def convert_birthweight(grams):
    pounds = grams * 0.00220462
    int_pounds = int(pounds) # Since pounds usually have decimals, decimals need to be converted into ounces
    ounces = (pounds - int_pounds) * 16
    return int_pounds, round(ounces,2)

# Example
birth_weight_in_grams = 2000 
int_pounds, ounces = convert_birthweight(birth_weight_in_grams)

print(f"Birth weight: {int_pounds} pounds and {ounces} ounces")


Birth weight: 4 pounds and 6.55 ounces


7. (5 pts) What is the average birth weight of babies whose mothers have a history of premature birth? 

In [16]:
premature_birth_data = data[ptl >= 1]

premature_birth_weights = premature_birth_data[:, 9]

avg_premature_birth_weight = round(np.mean(premature_birth_weights), 2)
print(f"Avg: {avg_premature_birth_weight} grams")


Avg: 2579.4 grams


8. (5 pts) What is the minimum, average and maximum birth weight of babies whose mother did not smoke during pregnancy and do not have a 
history of hypertension?

In [18]:
non_smoking_no_hypertension = data[(smoke == 0) & (hypertension == 0)]

non_smoking_no_hypertension_weights = non_smoking_no_hypertension[:, 9]

min_weight_ns_nh = round(np.min(non_smoking_no_hypertension_weights),2)
avg_weight_ns_nh = round(np.mean(non_smoking_no_hypertension_weights),2)
max_weight_ns_nh = round(np.max(non_smoking_no_hypertension_weights),2)

print(f"Min: {min_weight_ns_nh} grams, Avg: {avg_weight_ns_nh} grams, Max: {max_weight_ns_nh} grams")


Min: 1021.0 grams, Avg: 3090.44 grams, Max: 4990.0 grams


9. (5 pts) What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy?

In [20]:
smoking_mothers_data = data[smoke == 1]

smoking_mothers_weights = smoking_mothers_data[:, 9]

min_weight_smoking = round(np.min(smoking_mothers_weights), 2)
avg_weight_smoking = round(np.mean(smoking_mothers_weights), 2)
max_weight_smoking = round(np.max(smoking_mothers_weights), 2)

print(f"Min: {min_weight_smoking} grams, Avg: {avg_weight_smoking} grams, Max: {max_weight_smoking} grams")


Min: 709.0 grams, Avg: 2771.92 grams, Max: 4238.0 grams


10. (5 pts) What is the minimum, average and maximum birth weight of babies whose mothers had a history of hypertension? 

In [22]:
hypertensive_mothers_data = data[hypertension == 1]

hypertensive_mothers_weights = hypertensive_mothers_data[:, 9]

min_weight_ht = round(np.min(hypertensive_mothers_weights), 2)
avg_weight_ht = round(np.mean(hypertensive_mothers_weights), 2)
max_weight_ht = round(np.max(hypertensive_mothers_weights), 2)

print(f"Min: {min_weight_ht} grams, Avg: {avg_weight_ht} grams, Max: {max_weight_ht} grams")


Min: 1135.0 grams, Avg: 2536.83 grams, Max: 3790.0 grams


11. (5 pts) What is the minimum, average and maximum birth weight of babies whose mothers smoked during pregnancy and had a history of 
hypertension? 

In [24]:
smoking_and_hypertensive = data[(smoke == 1) & (hypertension == 1)]

smoking_and_hypertensive_weights = smoking_and_hypertensive[:, 9]

min_weight_sh = round(np.min(smoking_and_hypertensive_weights), 2)
avg_weight_sh = round(np.mean(smoking_and_hypertensive_weights), 2)
max_weight_sh = round(np.max(smoking_and_hypertensive_weights), 2)

print(f"Min: {min_weight_sh} grams, Avg: {avg_weight_sh} grams, Max: {max_weight_sh} grams")


Min: 1135.0 grams, Avg: 2561.0 grams, Max: 3756.0 grams


12. (5 pts) Compare and contrast the results from 7-11 above and summarize your findings.
Do you see any patterns in the birth weight of babies whose mother did not have any history of hypertension and/or did not smoke 
during pregnancy versus those who did? Ensure that you reference specific birth weights to support your explanation.


I have found that babies of non-smoking mothers without high blood pressure have the highest birth weights: these mothers' babies have the healthiest birth weights, with average birth weights usually higher than 3,000 grams, and maximum weights even reaching 5,000 grams.

On the contrary, the babies of mothers who smoked and suffered from high blood pressure had a lower birth weight: the average birth weight of the babies in the smoking group was lower, at 2771.92 grams, and the smallest weight of the babies in this group was only 709 grams, which shows the great influence of smoking on the development of the fetus. Similarly, the effect of high blood pressure on the development of the fetus was also very significant with the average weight of the babies in this group being 2536.83 grams.
Mothers who were both smokers and hypertensive had the lowest birth weight of their babies:
Although the mean birth weight of this group (2561.0 grams) was slightly higher than that of mothers with hypertension only, it was still much lower than that of mothers who were non-smokers and did not have hypertension.

Summary:
Both smoking and hypertension have a significant negative impact on infant birth weight, and the negative impact is more severe when the two risk factors are combined.
A history of preterm birth also affects infant birth weight, but the effect of preterm birth alone may not be as large as the superimposed effect of smoking and hypertension.

These results suggest that maternal health status, particularly smoking and hypertension, is directly related to infant birth weight, and that maintaining good health significantly improves infant birth weight.


## Helpful resources 
Don't forget to cite websites which helped you solve a problem in a unique way.  You can do this in markdown near the code or with a simple one-line comment inside the code cell, or you can list them below. 

You do not need to cite the official python documentation.
